### DISCUSSION TITLE
Dry run for bootstrap?

### DISCUSSION THREAD

**Discussion by ckeeney** [fluxcd/flux2#4795] [Proposals]
The bootstrap command looks great in the docs, but it's hard to run it for the first time with confidence in an existing cluster.  I wish there was a `--dry-run` flag that would do everything except stop at committing the files to the repository.

I have a single repository on Gitlab that has all my cluster manifests.  The repository has a build directory organized into 3 folders for 3 separate namespaces. Each namespace has its own Gitlab agent, and each Gitlab agent is watching only it's related namespace folder.

- production
- staging
- infrastructure (the `ingress-controller` and `sealed-secrets` and a few Kubernetes operators live in here)

I wanted to just try out Flux on the staging manifests, but the Gitlab docs and the Flux docs don't make it clear how to accomplish this.  My fear is that I run `flux bootstrap gitlab` and Flux makes unwanted changes to my production or infrastructure namespaces.

`flux bootstrap gitlab` does expose a `path` option. The [Flux docs](https://fluxcd.io/flux/cmd/flux_bootstrap_gitlab/#options) for this option sound promising:
> path relative to the repository root, when specified the cluster sync will be scoped to this path

But the [Gitlab docs](https://docs.gitlab.com/ee/user/clusters/agent/gitops/migrate_to_flux.html#bootstrap-flux) say 
> File path to a folder where the Flux configuration is stored.

Even if both of these docs are accurate, then it seems I will have Flux configuration files deployed in each of the 3 namespaces managed by my repository.  But that doesn't feel right to me.

Here is a different person having [issues with multiple Gitlab Agents interacting with Flux](https://forum.gitlab.com/t/gitops-flux-reconciliation-doesnt-work-properly-with-two-gitlab-agents/91268).  Admittedly his issue may be irrelevant, but it contributes to my hesitation to run the bootstrap command as a newcomer.

Ultimately, I wish there was a way for me to see what changes Flux wants to make without Flux making changes.  And by changes, I mean changes to the repository (bootstrapping could output the files and not commit them), and also to the cluster (bootstrapping could output the changes it intends to make to the cluster).

**Comment by souleb** [comment]
You may run `flux install --components-extra="image-reflector-controller,image-automation-controller" --export > test.yaml`. That will gives you all the manifests that will be installed so you can review them. `Flux bootstrap` also creates authentication secrets for syncing your cluster, so you won't have those. 

For dry-run, look at `flux diff kustomization` that can give you insight on what would be the change applied by a flux kustomization.
